File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/navigation/NavigationNotificationProvider.java
package com.mapbox.services.android.navigation.v5.navigation;

import android.content.Context;

import com.mapbox.services.android.navigation.v5.navigation.notification.NavigationNotification;
import com.mapbox.services.android.navigation.v5.routeprogress.RouteProgress;

class NavigationNotificationProvider {

  private NavigationNotification navigationNotification;
  private boolean shouldUpdate = true;

  NavigationNotificationProvider(Context applicationContext, MapboxNavigation mapboxNavigation) {
    navigationNotification = buildNotificationFrom(applicationContext, mapboxNavigation);
  }

  NavigationNotification retrieveNotification() {
    return navigationNotification;
  }

  void updateNavigationNotification(RouteProgress routeProgress) {
    if (shouldUpdate) {
      navigationNotification.updateNotification(routeProgress);
    }
  }

  void shutdown(Context applicationContext) {
    if (navigationNotification != null) {
      navigationNotification.onNavigationStopped(applicationContext);
    }
    navigationNotification = null;
    shouldUpdate = false;
  }

  private NavigationNotification buildNotificationFrom(Context applicationContext, MapboxNavigation mapboxNavigation) {
    MapboxNavigationOptions options = mapboxNavigation.options();
    if (options.navigationNotification() != null) {
      return options.navigationNotification();
    } else {
      return new MapboxNavigationNotification(applicationContext, mapboxNavigation);
    }
  }
}

File: libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/internal/navigation/metrics/NavigationNotificationProvider.kt
package com.mapbox.services.android.navigation.v5.internal.navigation.metrics

import android.content.Context
import com.mapbox.services.android.navigation.v5.internal.navigation.MapboxNavigationNotification
import com.mapbox.services.android.navigation.v5.navigation.MapboxNavigation
import com.mapbox.services.android.navigation.v5.navigation.notification.NavigationNotification
import com.mapbox.services.android.navigation.v5.routeprogress.RouteProgress

internal class NavigationNotificationProvider(
    applicationContext: Context,
    mapboxNavigation: MapboxNavigation
) {

    private val navigationNotification: NavigationNotification = buildNotificationFrom(applicationContext, mapboxNavigation)

    private var shouldUpdate: Boolean = true

    fun retrieveNotification(): NavigationNotification {
        return navigationNotification
    }

    fun updateNavigationNotification(routeProgress: RouteProgress) {
        if (shouldUpdate) {
            navigationNotification.updateNotification(routeProgress)
        }
    }

    fun shutdown(applicationContext: Context) {
        navigationNotification.onNavigationStopped(applicationContext)
        shouldUpdate = false
    }

    private fun buildNotificationFrom(
        applicationContext: Context,
        mapboxNavigation: MapboxNavigation
    ): NavigationNotification {
        val navNotification = mapboxNavigation.options().navigationNotification()
        return navNotification ?: MapboxNavigationNotification(applicationContext, mapboxNavigation)
    }
}
